File: mobile/src/main/java/net/mm2d/dmsexplorer/domain/model/ThemeModelImpl.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.domain.model;

import android.app.Activity;
import android.os.Build;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;

import net.mm2d.android.util.ActivityLifecycleCallbacksAdapter;
import net.mm2d.dmsexplorer.util.ThemeUtils;

import java.util.HashMap;
import java.util.Map;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public class ThemeModelImpl extends ActivityLifecycleCallbacksAdapter implements ThemeModel {
    private final Map<String, Integer> mMap = new HashMap<>();

    @Override
    public void setThemeColor(@NonNull final Activity activity,
                              @ColorInt final int toolbarColor,
                              @ColorInt final int statusBarColor) {
        final int color = statusBarColor != 0 ? statusBarColor : ThemeUtils.getDarkerColor(toolbarColor);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            activity.getWindow().setStatusBarColor(color);
        }
        mMap.put(activity.getClass().getName(), toolbarColor);
    }

    @Override
    public int getToolbarColor(@NonNull final Activity activity) {
        final Integer color = mMap.get(activity.getClass().getName());
        return color == null ? 0 : color;
    }

    @Override
    public void onActivityDestroyed(final Activity activity) {
        mMap.remove(activity.getClass().getName());
    }
}

File: mobile/src/main/java/net/mm2d/dmsexplorer/domain/model/ThemeModelImpl.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.domain.model

import android.app.Activity
import androidx.annotation.ColorInt
import net.mm2d.android.util.ActivityLifecycleCallbacksAdapter
import net.mm2d.dmsexplorer.util.ColorUtils

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
class ThemeModelImpl :
    ActivityLifecycleCallbacksAdapter(),
    ThemeModel {
    private val map = HashMap<String, Int>()

    override fun setThemeColor(
        activity: Activity,
        @ColorInt toolbarColor: Int,
        @ColorInt statusBarColor: Int,
    ) {
        val color = if (statusBarColor != 0) statusBarColor else ColorUtils.getDarkerColor(toolbarColor)
        activity.window.statusBarColor = color
        map[activity.javaClass.name] = toolbarColor
    }

    override fun getToolbarColor(activity: Activity): Int = map[activity.javaClass.name] ?: 0

    override fun onActivityDestroyed(activity: Activity) {
        map.remove(activity.javaClass.name)
    }
}
